### PR TITLE
Produce profiles in google trace format

### PR DESCRIPTION
**Pull request by SkySkimmer**
(trace format doc: https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/edit)

display using https://profiler.firefox.com/
or https://ui.perfetto.dev/
(or in chrome's builtin display but I didn't test it)

Example visualization (in perfetto):

- peano_naturals from HoTT (1MB gzipped)
peano_nats

- hurkens from stdlib (3MB uncompressed, 170KB gzipped)
hurkens

- example from https://github.com/coq/coq/issues/13977 (120MB uncompressed, 4MB gzipped)
bad

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
Bench failed to upload artifacts (too large)
I sshd into the machine, opam.NEW is 32GB, opam.OLD 21GB
Full list of trace files sorted by size https://gist.github.com/SkySkimmer/32721edc5ad8c70ed3fa49961e5f38b9

There are some approaches we could use to limit artifact size
- remove the files bigger than some size limit
- remove entries which takes less than some time limit (in the current streaming model this is inconvenient as we don't know how long it will take when we're printing the start event, and BTW "complete events" don't display as nicely IMO, but we could probably deal with it by delaying the printing) (they would still appear in the summary of subcalls for the surrounding events, and BTW we should add the number of calls next to the current sum of time spent)

Probably the second is more reasonable.

**Comment by SkySkimmer** [comment]
(that last fiat_crypto_via_setoid_rewrite_standalone is 19GB uncompressed BTW)

**Comment by Alizter** [comment]
Is .gz the best compression we can do?

**Comment by ejgallego** [comment]
As written in Zulip also note:
- https://github.com/coq/coq/pull/11791 (this PR may be a good time to remove cProfile)
-  https://github.com/ocaml/dune/issues/3449 
- https://github.com/coq/coq/issues/11542

**Comment by SkySkimmer** [comment]
I put some numbers in the "coq_makefile ..." commit
TLDR xz is somewhat better but not enough to compensate for being unable to give the compressed file directly to the visualizer IMO
Instead of 11GB we might have 8GB ish

**Comment by SkySkimmer** [comment]
Summary of `fiat_crypto_via_setoid_rewrite_standalone.v`:
~~~json
{
  "args": {
    "subtimes": {
      "Conversion": "5.33E+07 us, 12974960 calls",
      "Typeops.infer": "6.27E+04 us, 269 calls",
      "check-wf-univs": "3.45E+03 us, 269 calls",
      "close_proof": "3.47E+03 us, 76 calls",
      "command": "8.24E+08 us, 509 calls",
      "init": "5.45E+03 us, 1 call",
      "intern": "1.79E+05 us, 806 calls",
      "interp": "8.24E+08 us, 433 calls",
      "interp-delayed-qed": "2.63E+04 us, 76 calls",
      "kernel": "1.2E+05 us, 1377 calls",
      "parse_command": "2.87E+04 us, 510 calls",
      "pretyping": "1.43E+07 us, 3083964 calls",
      "synterp": "5.78E+04 us, 433 calls",
      "typeclass search": "1.2E+08 us, 2677 calls",
      "unfreeze_full_state": "1.01E+04 us, 736 calls",
      "unification": "2.45E+08 us, 67171072 calls",
      "universes_of_constr": "2.33E+03 us, 422 calls"
    },
    "major": "4.65E+08 w",
    "minor": "3.12E+11 w"
  },
  "name": "process",
  "ph": "E",
  "ts": 1686235869987325,
  "pid": 30626,
  "tid": 30626
}
~~~

**Comment by Janno** [comment]
This looks very interesting. I could see this becoming a less noisy alternative to `perf` when trying to answer high-level performance questions.

Do you know that printing directly won't lead to skewed results? When I implemented profiling in Mtac2 I had no choice but to collect everything into a list because the printing alone would distort nested events to the point of being unusable for any kind of microbenchmarks. It can be especially annoying when events have very different durations because the smallest events would be most affected by printing happening within them whereas for larger events it would almost make no difference, relatively speaking.

One thing that would improve the usefulness of this change for me would be to use `mtime`'s `to_uint64_ns` function (https://erratique.ch/software/mtime/doc/Mtime/index.html#val-to_uint64_ns) for timestamps. It corresponds to `MONOTONIC_CLOCK` on linux (which can be selected in `perf` with `perf record -k MONOTONIC_CLOCK`) and using it everywhere allows aligning traces across different profiling methods when they are collected simultaneously. We use it to align `perf` traces with our own internal automation traces and it works great. It gives us very fine-grained insight into which OCaml functions end up running during specific steps in our tactics.

**Comment by ejgallego** [comment]
IMHO a useful thing to have in a performance metric summary is a way to identify which calls are most costly, that is to say, "anomalous". For example it could be the case that the typechecking of some file is dominated by conversion or unification, but the interesting bit there is to try to know which calls are taking more time than they should.

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
Unrelated bench https://gitlab.com/coq/coq/-/jobs/4488727880: 126MB artifacts
This bench https://gitlab.com/coq/coq/-/jobs/4488304313: 193MB artifacts

**Comment by SkySkimmer** [comment]
Removed the minimum time based skip

@coqbot bench

**Comment by SkySkimmer** [comment]
>Unrelated bench gitlab.com/coq/coq/-/jobs/4488727880: 126MB artifacts
This bench gitlab.com/coq/coq/-/jobs/4488304313: 193MB artifacts

https://gitlab.com/coq/coq/-/jobs/4498167930: 240MB artifacts

**Comment by SkySkimmer** [comment]
The largest is for https://github.com/thery/coqprime/blob/master/src/Coqprime/examples/BasePrimes.v which is a file with 25k commands. The .gz is 1.3MiB.

File: doc/sphinx/practical-tools/coq-commands.rst
Comment (jfehrle) [inline doc/sphinx/practical-tools/coq-commands.rst:461]:
```suggestion
The output gives the duration and event counts for the execution of
components of Coq (for instance `process` for the whole file,
`command` for each command, `pretyping` for elaboration).
```

Maybe mention that some categories are included in others, eg command always includes parse_command time.

File: doc/sphinx/practical-tools/coq-commands.rst
Comment (jfehrle) [inline doc/sphinx/practical-tools/coq-commands.rst:463]: Hyperlink to the definition of the env variable

File: doc/sphinx/practical-tools/coq-commands.rst
Comment (jfehrle) [inline doc/sphinx/practical-tools/coq-commands.rst:464]:
```suggestion
which components produce events. This may be needed to reduce the
```

File: doc/sphinx/practical-tools/coq-commands.rst
Comment (jfehrle) [inline doc/sphinx/practical-tools/coq-commands.rst:465]:
```suggestion
size of the generated file.
```

File: doc/sphinx/practical-tools/coq-commands.rst
Comment (jfehrle) [inline doc/sphinx/practical-tools/coq-commands.rst:467]:
```suggestion
The generated file can be visualized with
```

File: doc/sphinx/practical-tools/coq-commands.rst
Comment (jfehrle) [inline doc/sphinx/practical-tools/coq-commands.rst:475]:
```suggestion
- `major` and `minor` indicate how many major and minor words were allocated during all the events.
```

From the example, it looks like these allocation .counts apply to all events under `args`

File: doc/sphinx/practical-tools/coq-commands.rst
Comment (jfehrle) [inline doc/sphinx/practical-tools/coq-commands.rst:477]: How about including a short sample output file illustrating the various fields?

File: doc/sphinx/practical-tools/coq-commands.rst
Comment (jfehrle) [inline doc/sphinx/practical-tools/coq-commands.rst:478]:
```suggestion
  how many times each subcomponent was profiled during the event
```

File: doc/sphinx/practical-tools/coq-commands.rst
Comment (jfehrle) [inline doc/sphinx/practical-tools/coq-commands.rst:482]: Would be great to show sample output for this.  Sounds like there is one `cmd` per unique command name.  Right?

File: doc/sphinx/practical-tools/utilities.rst
Comment (jfehrle) [inline doc/sphinx/practical-tools/utilities.rst:740]:
```suggestion
timing data: per-file build-times, per-line times for individual
files, and profiling data in Google trace format for individual
files.
```

**Review by jfehrle — COMMENTED** [review]
wording suggestions

File: tools/CoqMakefile.in
Comment (gares) [inline tools/CoqMakefile.in:807]: @SkySkimmer isn't there an extra dollar?

**Review by gares — APPROVED** [review]
makefile changes look ok to me

**Comment by gares** [comment]
@SkySkimmer this PR looks almost ready. Can you take care of the doc suggestions quickly?

File: tools/CoqMakefile.in
Comment (SkySkimmer) [reply tools/CoqMakefile.in:807]: No, this is like TIMER and TIMING_ARG, evaluation needs to be delayed as when globvorule is expanded the `$<` is not yet defined

File: doc/sphinx/practical-tools/coq-commands.rst
Comment (SkySkimmer) [reply doc/sphinx/practical-tools/coq-commands.rst:174]: This is how the other refs around this part are done

File: doc/sphinx/practical-tools/coq-commands.rst
Comment (SkySkimmer) [reply doc/sphinx/practical-tools/coq-commands.rst:182]: >Would be nice to include the current list of component names.

It would not be nice to have to maintain it.

File: doc/sphinx/practical-tools/coq-commands.rst
Comment (SkySkimmer) [reply doc/sphinx/practical-tools/coq-commands.rst:461]: >Maybe mention that some categories are included in others, eg command always includes parse_command time.

This could change in the future.

File: doc/sphinx/practical-tools/coq-commands.rst
Comment (SkySkimmer) [reply doc/sphinx/practical-tools/coq-commands.rst:475]: They are annotated on a specific event and correspond to that event. The presence of subevents is irrelevant.

File: doc/sphinx/practical-tools/coq-commands.rst
Comment (SkySkimmer) [reply doc/sphinx/practical-tools/coq-commands.rst:482]: I don't know what you mean by "unique command name", `cmd` looks like `Chars 54 - 72 [Definition~a~:=~1.]`

**Comment by SkySkimmer** [comment]
@coqbot run full ci

should be done

**Comment by SkySkimmer** [comment]
Do we want to delete cprofile and renamed newprofile -> cprofile?

**Comment by ejgallego** [comment]
> Do we want to delete cprofile and renamed newprofile -> cprofile?

Yes. Or even maybe `CTrace`, or even better we could use `(wrapped true)` and stop the `C` mess.

**Review by ejgallego — CHANGES_REQUESTED** [review]
Patch look good to me, however please don't embed a "minijson" library in the new profile file; let's try to consolidate json use in Coq.

The same can be said about the trace format, there are OCaml libs doing that already (like the one in Dune and I believe in base), please let's separate json, catapult, and the profiler into 3 modules.

The trace format is not called "Google Trace Format" but I think it is called "Catapult" (not so sure about this tho, googling is confusing about this)

**Comment by SkySkimmer** [comment]
>The same can be said about the trace format, there are OCaml libs doing that already (like the one in Dune and I believe in base), please let's separate json, catapult, and the profiler into 3 modules.

That sounds overcomplicated for what we do with it.

>The trace format is not called "Google Trace Format" but I think it is called "Catapult" (not so sure about this tho, googling is confusing about this)

The doc I found calls it "trace event format" so that's what I use.

**Comment by ejgallego** [comment]
> That sounds overcomplicated for what we do with it.

Feel free to dismiss my request then, however we often have this discussion about using external dependencies vs our own do it yourself apporach which I have to say has many downsides vs relying on mature, stable deps.

For example, in this particular case, the current setup is a PITA to use from coq-lsp.

Ideally, what I would like coq-lsp to do here is when the user runs the command "trace my document", I can get the results of the trace in a standard `Yojson.Safe.t` type, so I can then just embed them in a message and have the client display them.

With the apporach taken here, I have either to unnecesarily write a file and parse the JSON , or to convert from the ad-hoc format to the regular one. Moreover, it becomes very hard for me to adapt the tracing system to it becomes compatible with incremental checking.

There are other similar assumptions in the API (such as the `at_exit`) that makes this API very hard to use for what I want.

So I'd recommend we try to find a quick balance so this PR is not blocked, but such a use case above is possible.

[I'm happy to implement a prototype "profile file" in coq-lsp if you are curious]

File: lib/newProfile.ml
Comment (ejgallego) [inline lib/newProfile.ml:200]: Please make this explicit, otherwise this can't work with `coq-lsp` and other systems that handle multilpe documents.

**Review by ejgallego — CHANGES_REQUESTED** [review]
I think that `finish` needs to be exposed, and `accu` and `sums` need to be registered with the summary, otherwise this is not usable from `coq-lsp` and (likely other DMs)

An alternative to the summary is exposing an API to restore / save the profile state, that would work very with incremental document processing.

**Comment by SkySkimmer** [comment]
It should for sure not use the summary.

Anyway it's intended for use by coqc, interactive use has not been considered.

**Comment by ejgallego** [comment]
> It should for sure not use the summary.

I was thinking that would work fine, but then just a save/restore API plust a `CProfile.t` type that stores the state would work fine.
 
> Anyway it's intended for use by coqc, interactive use has not been considered.

Even for coqc the `at_exit` could become problematic at some point, IMHO it is best to handle that explicitly.

These are changes we want to do for sure, I guess it is best to make them here.

**Comment by rlepigre** [comment]
> Anyway it's intended for use by coqc, interactive use has not been considered.

Indeed, for us the `coqc` use-case is what we care about, combination with https://github.com/coq/coq/pull/17744 which is currently block until this PR is merged.

I don't see any good interactive use-case for the generation of profiling data. What we typically want to do with such data is generate it automatically in CI, and then also automatically generate some for of report.

Do you have any specific use-case in mind @ejgallego?

**Comment by ejgallego** [comment]
> Do you have any specific use-case in mind @ejgallego?

Displaying this data in UIs, like the coq-lsp VSCode plugin (which already has a performance view)

Also, in its current form, the patch won't work with other compilers like `fcc` due to `at_exit`.

I think the change requested to be able to handle the per-file stats properly is pretty small IMHO.

**Comment by SkySkimmer** [comment]
@ejgallego can we unblock this?
I don't have a great understanding of the requested changes, so IMO we need to either have a visio to do them together, or merge now and do them later.

>These are changes we want to do for sure, I guess it is best to make them here.

Indeed I disagree about this, I don't think there's anything preventing us from making changes after merging.

**Comment by ejgallego** [comment]
> @ejgallego can we unblock this?

Yes, let's unblock this.

However I'm still concerned of using our own custom JSON type, I think I'd prefer we vendor a small part of an established library. Should we discuss this in Zulip / a call? Having to reinvent the wheel everytime Coq needs something standard doesn't seem like the best approach to me.

> I don't have a great understanding of the requested changes, so IMO we need to either have a visio to do them together, or merge now and do them later.

Let's do these changes later. But basically what I need is:
- a way to restore / set profile state, or a way to do it per-sentence / API call
- a way to obtain this data in a buffered way instead of writing to a file.

> > These are changes we want to do for sure, I guess it is best to make them here.
> 
> Indeed I disagree about this, I don't think there's anything preventing us from making changes after merging.

I agree.

**Comment by ejgallego** [comment]
@SkySkimmer note the ppx_tracing and tracing lib Jane Street has been developing (c.f. https://blog.janestreet.com/what-the-interns-have-wrought-2023/ )

File: lib/newProfile.ml
Comment (rlepigre) [inline lib/newProfile.ml:34]: What I've done in the past in other projects when I needed JSON is use polymorphic variants that are compatible with those of yojson. For instance, here you could do:
```suggestion
  type t = [ 
    | `Intlit of string
    | `String of string
    | `Assoc of (string * t) list
    | `List of t list
  ]
```
which would be compatible with `Yojson.Safe.t` via subtyping I believe.

Maybe using a type like this (which is isomorphic to yours) would suit @ejgallego better.

Comment (ejgallego) [reply lib/newProfile.ml:34]: Indeed, tho I'd be still concerned about having to deploy our own utils for pretty printing, etc...

IMHO we need to have the discussion about using external deps in the project. I can understand that it is frustating to have that discussion on PRs where this issue is just tangential, and indeed it seems not a good choice to block this PR on that discussion.

However on the other hand, due to postponing the discussion, we keep adding technical debt to Coq, or limiting our options as developers, so each time we go ahead and add things like this we save some time, at the cost of maybe a lot of time in the future. Delaying also has a lot of cost. I dunno how to do best.

In this case note that we already have another JSON type in the extraction plugin, right?

Comment (ejgallego) [reply lib/newProfile.ml:34]: So apart from the external deps discussion that honestly I'm unsure how to have, I think @rlepigre suggestion is good actually, pretty easy to do, and means that external tools that already use `Yojson.Safe.t` can interop fine with this, so maybe it is worth tweaking it before merge.

Comment (SkySkimmer) [reply lib/newProfile.ml:34]: I'm fine with that suggestion.

I do like having control of the printing here because it means we can print 1 event / line, which means we can use grep to do a bit of manual analysis when the data gets big enough that smarter tools fail (specifically I tried jq on one of the big files from the earlier benches and it just kept allocating IIRC)

Comment (ejgallego) [reply lib/newProfile.ml]: That's fine, in a later PR we need to have the print to happen to a buffer (not a file) tho, so the LSP client can relay this info without having to do a round-trip by the FS. The buffer may get big indeed, but it is assumed that for this kind of tracing you need quite a beefy machine anyways.

**Review by ejgallego — APPROVED** [review]
We will address the API tweaks in a further PR.

The issue of what JSON library Coq should use is referred to the next call.

**Comment by SkySkimmer** [comment]
I considered printing to a buffer originally but using Printf.fprintf in the subprints forces out_channel which confused me.
I guess it should be fine to use Format instead, then we can do of_out_channel of_buffer etc.

**Comment by ejgallego** [comment]
> I considered printing to a buffer originally but using Printf.fprintf in the subprints forces out_channel which confused me. I guess it should be fine to use Format instead, then we can do of_out_channel of_buffer etc.

Yes. IMHO the current default is good for coqc, it is better to flush directly to disk than to take valious RAM. Finer printing is only needed when you want to recheck only a part of the document.

**Comment by SkySkimmer** [comment]
I think I understand better what you want after thinking about this printf / format stuff, cf new commits

**Comment by ejgallego** [comment]
> I think I understand better what you want after thinking about this printf / format stuff, cf new commits

Thanks Gaëtan, this will be a good basis to start implementing the incremental support.

I'm happy to merge now, not sure if we want to run the full CI tho? Gonna do it "just in case"

I'd be nice to have the `dune build @check` for all plugins in the light pipeline, that would help quite a bit in PRs like this.

**Comment by ejgallego** [comment]
@coqbot run full CI

File: lib/newProfile.ml
Comment (ejgallego) [inline lib/newProfile.ml:66]: @SkySkimmer you may want to add a flush here if you want to have fully synchonous semantics, `Format` can buffer aggresively some times.

This can also be done using the specifier `%!` , or via the `kprintf` family of functions.

**Comment by ejgallego** [comment]
@coqbot: merge now

**Comment by ejgallego** [comment]
Still note the comments about flush, that could need a tweak in a further PR.

**Pull request merged**
